### PR TITLE
Use current revisions and don't unpublish anything!

### PR DESCRIPTION
**Pull request by jkopel**
Modify the handling of revisions so that any import makes a new version of the target but leaves the current published node alone if there is a draft. Also should fix issues where targets get unpublished when there is a draft over a published version.
Also tests!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -65,8 +65,13 @@ before_script:
 
 script:
   - phpunit --coverage-text --coverage-clover build/logs/clover.xml
-  - if [ $BDD = 'true' ] ; then behat --suite=contentTranslation --verbose ; fi
   - if [ $BDD = 'true' ] ; then behat --suite=entityTranslation --verbose ; fi
+  # Disable entity_translation to run content translation tests.
+  - if [ $BDD = 'true' ] ; then pushd build/drupal ; fi
```
Comment (iamEAP) [inline .travis.yml]: - [ ] Consider collapsing into one line?

```yaml
  - if [ $BDD = 'true' ] ; then pushd build/drupal && drush dis entity_translation --y && popd ; fi
```

File: modules/workbench_moderation.entity_xliff.inc [hunk 2]
```diff
@@ -18,17 +18,24 @@ if (!function_exists('workbench_moderation_entity_xliff_translatable_source_alte
...
         // Then wrap the current revision, not the live version.
-        $node = workbench_moderation_node_current_load($wrapper->value());
-        // If this has not been saved yet it will return false, in which case just use the wrapper.
-        if ($node) {
-          $wrapper->set($node);
+        $current_node = workbench_moderation_node_current_load($wrapper->value());
+        $published_node = node_load($id);
+        if ($current_node->vid != $published_node->vid) {
+          // There is a published node and a current Draft.
+          // First force the published node to stay published
```
Comment (iamEAP) [inline modules/workbench_moderation.entity_xliff.inc]: - [ ] Is the node loaded by `node_load()` necessarily published?  Is it possible that the node had never been published and this would be publishing it?

File: modules/workbench_moderation.entity_xliff.inc [hunk 3]
```diff
@@ -61,3 +70,47 @@ if (!function_exists('workbench_moderation_entity_xliff_target_entities_alter'))
...
+      if (!empty($id)) {
+        $current_node = workbench_moderation_node_current_load($node);
+        $published_node = node_load($node->nid);
+        if ($current_node->vid != $published_node->vid) {
+          // There is a published node and a current Draft.
+          $wrapper->save();
+          $wrapper->set($published_node);
+          $node->workbench_moderation_state_new = workbench_moderation_state_none();
+        }
+      }else{
```
Comment (iamEAP) [inline modules/workbench_moderation.entity_xliff.inc]: - Nit: `else` should be on newline w/space before the trailing brace.

File: src/Drupal/Translatable/Content/NodeTranslatable.php [hunk 4]
```diff
@@ -113,7 +112,8 @@ public function getTargetEntity($targetLanguage) {
 
         // Do not mark this node as a new revision. This is necessary in
         // cases where this node happens to reference a field collection...
-        $target->revision = FALSE;
```
Comment (iamEAP) [inline src/Drupal/Translatable/Content/NodeTranslatable.php]: - [ ] Probably need to update the comment, here.

File: test/Drupal/Translatable/Content/FieldCollectionTranslatableTest.php [hunk 5]
```diff
@@ -231,28 +231,153 @@ public function getTargetEntityFromStaticCache() {
...
+      $this->assertFalse(isset($actualTarget->item_id));
+      $this->assertFalse(isset($actualTarget->revision_id));
+      $this->assertEquals(TRUE, $actualTarget->is_new);
+    }
 
-      $observerWrapper = $this->getMock('\EntityDrupalWrapper', array('value'));
-      $observerWrapper->expects($this->once())
+    /**
+     * Tests the getTargetEntity method in the case that the host language does
+     * not match the target, and thus we need to create brand new field collection
```
Comment (iamEAP) [inline test/Drupal/Translatable/Content/FieldCollectionTranslatableTest.php:319]: - [x] Do you mean, "in the case that the host language _does_ match the target?"

File: test/Features/Bootstrap/FeatureContext.php [hunk 6]
```diff
@@ -56,8 +56,36 @@ public function gatherContexts(BeforeScenarioScope $scope) {
...
+`  -----' `-----  '
+
+END;
     die();
   }
+
+  /**
+   * @Then I want to debug this :node
+   */
+  public function iWantToDebugThisNode($node)
```
Comment (iamEAP) [inline test/Features/Bootstrap/FeatureContext.php]: - [ ] One thing I found while I was working on Trust yesterday (which also uses Behat) is that this functionality already exists in the main DrupalContext:

https://github.com/jhedstrom/drupalextension/blob/master/src/Drupal/DrupalExtension/Context/DrupalContext.php#L426

File: test/Features/contentTranslation/AlternateSourceLanguage.feature [hunk 7]
```diff
@@ -29,19 +30,21 @@ Feature: Alternative Source Languages
...
     # Re-import to test the pre-existing/non-initialization flow.
     When I click "Translate"
     And I click "French page title"
     And I click "XLIFF"
     And I attach an "en" translation of this "French" node
     And I press the "Import" button
     Then I should see the success message containing "Successfully imported"
-
     And there should be no corrupt translation sets.
+    When I click "Translate"
```
Comment (iamEAP) [inline test/Features/contentTranslation/AlternateSourceLanguage.feature]: - [ ] Necessary?

File: src/Drupal/Translatable/Content/FieldCollectionTranslatable.php [hunk 8]
```diff
@@ -51,8 +51,33 @@ public function getSourceLanguage() {
    */
   public function getTargetEntity($targetLanguage) {
     if (!isset($this->targetEntities[$targetLanguage]) || empty($this->targetEntities[$targetLanguage])) {
+      $target = $this->getRawEntity($this->entity);
```
Comment (iamEAP) [inline src/Drupal/Translatable/Content/FieldCollectionTranslatable.php:54]: - [x] Are these changes necessary, or can we avoid making changes here and punt until we/someone do #134?

File: test/Features/contentTranslation/NodeReferences.feature [hunk 9]
```diff
@@ -7,10 +7,10 @@ Feature: Node References
...
     Given I am logged in as a user with the "administer entity xliff" permission
     And "page" content:
-      | title                              | field_long_text                    | promote |
-      | English node host page title       | English node host body text.       | 1       |
-      | English node child page title      | English node child body text.      | 1       |
-      | English node grandchild page title | English node grandchild body text. | 1       |
+      | title                              | field_long_text                    | promote | status |
+      | English node host page title       | English node host body text.       | 1       | 1       |
+      | English node child page title      | English node child body text.      | 1       | 1       |
+      | English node grandchild page title | English node grandchild body text. | 1       | 1       |
```
Comment (iamEAP) [inline test/Features/contentTranslation/NodeReferences.feature:13]: - [x] Any reason these are okay, but other content requires the _through-the-UI_ approach to test set up?

File: test/scripts/install_drupal.sh [hunk 10]
```diff
@@ -41,9 +41,11 @@ fi
 # Place this module into the sites/all/modules directory and enable it.
 rsync -aq "`pwd`" "`pwd`/$BUILD_DIR/drupal/sites/all/modules/entity_xliff" --exclude build
 # pin entity-7.x-1.7 until this gets fixed: https://www.drupal.org/node/2807275
+# Pin references-7.x-2.1 since the module is unsupported do to unresolved security issues.
```
Comment (iamEAP) [inline test/scripts/install_drupal.sh]: - Nit: _due_ to

**Review by iamEAP — CHANGES_REQUESTED** [review]

File: test/Drupal/Translatable/Content/FieldCollectionTranslatableTest.php (hunk 5 above)
Comment (jkopel) [reply test/Drupal/Translatable/Content/FieldCollectionTranslatableTest.php:319]: Not sure, that is a revert to a test you wrote. :)

File: test/Features/Bootstrap/FeatureContext.php (hunk 6 above)
Comment (jkopel) [reply test/Features/Bootstrap/FeatureContext.php]: Whoa! I will kill this then.

File: test/Features/contentTranslation/NodeReferences.feature (hunk 9 above)
Comment (jkopel) [reply test/Features/contentTranslation/NodeReferences.feature:13]: It has to do with the expectation that workbench will have completed whatever it needs to do. In this case we are not looking at the published status AFTER doing some other operation. I only changed the ones that really had to change since the "And "page" content:" version is so much more readable.

File: .travis.yml (hunk 1 above)
Comment (jkopel) [reply .travis.yml]: Is this better?
I rather prefer reading it line by line.

File: modules/workbench_moderation.entity_xliff.inc (hunk 2 above)
Comment (jkopel) [reply modules/workbench_moderation.entity_xliff.inc]: If that is the case then the "published" version still matches the "current" version which would BOTH be the draft.
I could certainly change the variable name for clarity.

File: src/Drupal/Translatable/Content/FieldCollectionTranslatable.php (hunk 8 above)
Comment (jkopel) [reply src/Drupal/Translatable/Content/FieldCollectionTranslatable.php:54]: These changes are a reversion to a previous state. 
I got confused by the "field is translatable" setting causing all kinds of havoc.
If I don't revert these then I need to comment out a whole lot of other tests...

File: src/Drupal/Translatable/Content/NodeTranslatable.php (hunk 4 above)
Comment (jkopel) [reply src/Drupal/Translatable/Content/NodeTranslatable.php]: Rats, thought I had but it must have been in the build version.

File: .travis.yml (hunk 1 above)
Comment (jkopel) [reply .travis.yml]: I guess it would be marginally faster. Will change.

File: modules/workbench_moderation.entity_xliff.inc (hunk 2 above)
Comment (iamEAP) [reply modules/workbench_moderation.entity_xliff.inc]: I guess the concern is that it's being explicitly set to `published` below.

Comment (iamEAP) [reply modules/workbench_moderation.entity_xliff.inc]: Aaaah.  I see.  This condition never fires if you have a node has never been published.  As discussed: some clarity (as comments) around that would be helpful.

File: test/Drupal/Translatable/Content/FieldCollectionTranslatableTest.php (hunk 5 above)
Comment (jkopel) [reply test/Drupal/Translatable/Content/FieldCollectionTranslatableTest.php:319]: This is the place where we are having trouble. The test is correct, since the code IS trying to do something different when the host language != the target language. This is where we have a listwrapper and the "host" is ending up looking at the source instead of the target.

File: test/scripts/install_drupal.sh (hunk 10 above)
Comment (jkopel) [reply test/scripts/install_drupal.sh]: (doh)

**Comment by iamEAP** [comment]
Thanks @jkopel! Merging and pushing up to d.o.

**Review by iamEAP — APPROVED** [review]
